README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# go2cell

<!-- badges: start -->

<!-- badges: end -->

By using [Wikidata](https://www.wikidata.org/wiki/Wikidata:Main_Page)
and a local RDF database of cell type gene markers, go2cell links Gene
Ontology terms to cell types, therefore enriching Omics results with
open semantic knowledge.

## Installation

You can install the latest version of go2cell from GitHub with:

``` r
install.packages("remotes")
remotes::install_github("jvfe/go2cell")
```

## Quickstart

Suppose we want to retrieve cell types with gene markers related to GO
terms ‘muscle contraction’ and ‘glycoprotein metabolic process’:

``` r
library(go2cell)

go_ids <- c("GO:0006936", "GO:0009100")

results <- go2cell(go_ids)
```

| cell\_type    | cell\_typeLabel     | go\_ids      | go\_termLabel                  | geneLabel |
| :------------ | :------------------ | :----------- | :----------------------------- | :-------- |
| wd:Q101404949 | human myofibroblast | <GO:0006936> | muscle contraction             | CALD1     |
| wd:Q101404949 | human myofibroblast | <GO:0006936> | muscle contraction             | ACTA2     |
| wd:Q101404949 | human myofibroblast | <GO:0006936> | muscle contraction             | MYL9      |
| wd:Q101404949 | human myofibroblast | <GO:0006936> | muscle contraction             | DES       |
| wd:Q101404922 | human epsilon cell  | <GO:0009100> | glycoprotein metabolic process | PCSK6     |
| wd:Q101404940 | human myoblast      | <GO:0006936> | muscle contraction             | TPM2      |


NEWS.md
# go2cell 0.99.0

* Added all basic function logic, as well as tests and documentation.



tests/testthat.R
library(testthat)
library(go2cell)

test_check("go2cell")


tests/testthat/test-main.R
go_ids <- c(
  "GO:0006936", "GO:0008152", "GO:0015031",
  "GO:0006468", "GO:0010043", "GO:0006837",
  "GO:0006359", "GO:0005978", "GO:0007260", "GO:0014873"
)

celltype_qids <- c("Q101404881", "Q101404901", "Q101404922")

broken_id <- "GO_0006936"
# broken_qid <-

test_that("go2cell works", {
  celltype_info <- go2cell(go_ids)

  expect_setequal(unique(celltype_info$go_ids), go_ids)
  expect_true(nrow(celltype_info) > 100)
  expect_equal(ncol(celltype_info), 5)
})

test_that("cell2go works", {
  celltype_info <- cell2go(celltype_qids)

  expect_setequal(unique(celltype_info$cell_type), stringr::str_glue("wd:{celltype_qids}"))
  expect_true(nrow(celltype_info) > 3400)
  expect_equal(ncol(celltype_info), 5)
})

test_that("broken GO ids raise error", {
  expect_error(go2cell(broken_id))
})


tests/testthat/test-utils.R
to_collapse <- c(rep(c("GO:0006936", "GO:0008152"), 7))
to_collapse_single <- "GO:0006936"

to_clean <- c("http://www.wikidata.org/entity/Q18031649", "http://www.wikidata.org/entity/Q227339")

test_that("basic collapsing works", {
  vec_collapsed <- .collapse_as_values(to_collapse)
  single_collapsed <- .collapse_as_values(to_collapse_single)

  expect_equal(vec_collapsed, "(GO:0006936) (GO:0008152)")
  expect_equal(single_collapsed, "(GO:0006936)")
})

test_that("quoted collapsing works", {
  vec_collapsed <- .collapse_as_values(to_collapse, quotes = TRUE)
  single_collapsed <- .collapse_as_values(to_collapse_single, quotes = TRUE)

  expect_equal(vec_collapsed, "('GO:0006936') ('GO:0008152')")
  expect_equal(single_collapsed, "('GO:0006936')")
})

test_that("URL removal works", {
  clean <- .remove_wdt_url(to_clean)

  expect_equal(clean, c("wd:Q18031649", "wd:Q227339"))
})


R/utils.R
#' Format string or character vector as SPARQL-compatible list
#'
#' @param item_vector A string or a character vector of items
#' @param quotes A logical to indicate whether or not to encase in quotes
#'
#' @return A SPARQL-compatible string of the values
#' @examples
#' .collapse_as_values(c("GO:0006936", "GO:0008152"))
.collapse_as_values <- function(item_vector, quotes = FALSE) {
  unique_items <- stringr::str_trim(unique(item_vector))

  if (quotes) {
    paste(paste0("('", unique_items, "')"), collapse = " ")
  } else {
    paste(paste0("(", unique_items, ")"), collapse = " ")
  }
}

#' Substitute the wikidata url for a SPARQL-friendly wd:
#'
#' @param column The column of the dataframe to substitute
#'
#' @return A character vector
#' @examples
#' .remove_wdt_url("http://www.wikidata.org/entity/Q18031649")
.remove_wdt_url <- function(column) {
  stringr::str_replace_all(
    column, "http://www.wikidata.org/entity/",
    "wd:"
  )
}


R/go2cell.R
#' Get cell types from GO ids
#'
#' Return cell type information for specific Gene Ontology identifiers
#'
#' This function takes in a string or a character vector of
#' Gene Ontology (GO) identifiers, queries Wikidata
#' (\url{https://www.wikidata.org/})
#' for genes related to said identifiers, and crosses this data
#' with a local RDF database of cell type-marker correspondence,
#' returning, at last, a dataframe of cell type information related to
#' the given GO identifiers.
#'
#' @param go_ids A string or a character vector of Gene Ontology identifiers.
#'
#' @return A dataframe of 7 columns: The first two correspond to cell types,
#'    that is, their Wikidata Identifier and their name. The two following
#'    columns correspond to the Gene Ontology IDs given, and the last column
#'    corresponds to the cell type's marker that led to the result.
#'
#' @seealso \url{https://www.wikidata.org/} for information regarding
#'   Wikidata and \url{https://panglaodb.se/index.html} for the
#'   original database this information was adapted from.
#'
#' @export
#'
#' @examples
#' # Return cell types related to 'muscle contraction' and 'metabolism'
#' go2cell(c('GO:0006936', 'GO:0008152'))
#'
#' # Return cell types related to 'response to zinc ion'
#' go2cell('GO:0010043')
#' \dontrun{
#' # IDs should use ':' not '_'
#' go2cell('GO_0010043')
#' }
go2cell <- function(go_ids) {
    go_ids_collapsed <- .collapse_as_values(go_ids, quotes = TRUE)

    genes_from_go <- .get_genes_from_go(go_ids_collapsed)

    gene_values <- .collapse_as_values(genes_from_go$gene)

    celltypes <- .query_ctp_turtle(gene_values, query_key = "cell_type", selector = "gene") %>%
        dplyr::filter(gene %in% genes_from_go$gene)

    celltypes_collapsed <- .collapse_as_values(celltypes$cell_type)

    celltype_wdt <- .get_celltype_items(celltypes_collapsed)

    final_table <- celltypes %>% dplyr::inner_join(genes_from_go,
        by = "gene") %>% dplyr::left_join(celltype_wdt, by = c("cell_type")) %>%
        dplyr::select(cell_type, cell_typeLabel, go_ids, go_termLabel,
            geneLabel)
}


R/queries.R
#' Query Wikidata for genes related to GO ids
#'
#' @param go_ids A character vector of GO identifiers
#'
#' @return A dataframe of GO ids and genes
#' @examples
#' .get_genes_from_go("('GO:0006936')")
.get_genes_from_go <- function(go_ids) {
  if (any(grepl("GO:\\d+", go_ids) == FALSE)) {
    stop("A GO identifier given doesn't match the regex 'GO:\\d+',
             check your input for inconsistencies.")
  }

  message("Querying wikidata for GO/gene correspondence...")
  go_query <- sprintf(
    "SELECT ?gene ?go_ids ?go_termLabel ?geneLabel
      WHERE
      {
        VALUES (?go_ids) {%s}
        ?go_term wdt:P686 ?go_ids.
        ?protein wdt:P703 wd:Q15978631;
                 ?godomain ?go_term;
                 wdt:P702 ?gene.
        SERVICE wikibase:label { bd:serviceParam wikibase:language \"en\". }
      }",
    go_ids
  )

  WikidataQueryServiceR::query_wikidata(go_query) %>%
    dplyr::mutate(gene = .remove_wdt_url(gene))
}

#' Query local database for cell types related to particular markers
#'
#' @param query_values A character vector of SPARQL-valid values
#' @param query_key A character of the key value to select items by
#' @param selector A character vector of the query_values entity
#' @param mode A character indicating whether to run go2cell or cell2go
#'
#' @return A dataframe of cell type Wikidata items and their respective markers
#' @examples
#' .query_ctp_turtle("wd:Q18016342", "cell_type", "gene", mode="go2cell")
.query_ctp_turtle <- function(query_values, query_key, selector, mode) {
  celltype_marker_graph <- rdflib::rdf_parse(celltype_marker_turtle,
    format = c("turtle")
  )

  if (mode == "go2cell") {
    ctp_sparql <- stringr::str_glue(
      "PREFIX ctp: <http://celltypes.wiki.opencura.com/entity/>
     PREFIX wd: <http://www.wikidata.org/entity/>
      SELECT ?{query_key} ?{selector}
           WHERE {{
            VALUES (?{selector}) {{({query_values})}}
            ?{query_key} ctp:P9 {query_values}.
           }}"
    )
  } else {
    ctp_sparql <- stringr::str_glue(
      "PREFIX ctp: <http://celltypes.wiki.opencura.com/entity/>
     PREFIX wd: <http://www.wikidata.org/entity/>
      SELECT ?{query_key} ?{selector}
           WHERE {{
            VALUES (?{selector}) {{({query_values})}}
            {query_values} ctp:P9 ?{query_key}.
           }}"
    )
  }


  lapply(ctp_sparql, rdflib::rdf_query, rdf = celltype_marker_graph) %>%
    dplyr::bind_rows() %>%
    dplyr::mutate(
      cell_type = .remove_wdt_url(cell_type),
      gene = .remove_wdt_url(gene)
    )
}

#' Query Wikidata for cell type item names from cell type QIDs
#'
#' @param celltype_ids A character vector of cell type QIDs
#'
#' @return A dataframe of QIDs and their respective labels
#' @examples
#' .get_celltype_items("(wd:Q101405098) (wd:Q101404862)")
.get_celltype_items <- function(celltype_ids) {
  message("Querying wikidata for cell type labels...")
  celltype_sparql <- sprintf(
    "SELECT ?cell_type ?cell_typeLabel
  WHERE
  {
    VALUES (?cell_type) {%s}
    SERVICE wikibase:label { bd:serviceParam wikibase:language \"en\". }
  }",
    celltype_ids
  )

  celltype_wdt <- WikidataQueryServiceR::query_wikidata(celltype_sparql) %>%
    dplyr::mutate(cell_type = .remove_wdt_url(cell_type))
}

#' Query Wikidata for GO ids associated with particular genes
#'
#' @param genes A character vector of SPARQL-valid gene items
#'
#' @return A dataframe of GO ids and genes
#'
#' @examples
#' .get_go_from_genes("(wd:Q15317282) (wd:Q14912592) (wd:Q14912176)")
.get_go_from_genes <- function(genes) {
  message("Querying wikidata for GO/gene correspondence...")
  gene_sparql <- stringr::str_glue(
    "SELECT ?gene ?go_term ?go_itemLabel ?geneLabel
      WHERE
      {{
        VALUES (?gene) {{{genes}}}
        ?gene wdt:P703 wd:Q15978631;
              wdt:P688 ?protein .
        ?protein (wdt:P680|wdt:P681|wdt:P682) ?go_item.
        ?go_item wdt:P686 ?go_term.

        SERVICE wikibase:label {{ bd:serviceParam wikibase:language 'en'. }}
      }}"
  )

  WikidataQueryServiceR::query_wikidata(gene_sparql) %>%
    dplyr::mutate(gene = .remove_wdt_url(gene))
}


R/main.R
#' Get cell types from GO ids
#'
#' Return cell type information for specific Gene Ontology identifiers
#'
#' This function takes in a character vector of
#' Gene Ontology (GO) identifiers, queries Wikidata
#' (\url{https://www.wikidata.org/})
#' for genes related to said identifiers, and crosses this data
#' with a local RDF database of cell type-marker correspondence,
#' returning, at last, a dataframe of cell type information related to
#' the given GO identifiers.
#'
#' @param go_ids A character vector of Gene Ontology identifiers.
#'
#' @return A dataframe of 5 columns: The first two correspond to cell types,
#'    that is, their Wikidata Identifier and their name. The two following
#'    columns correspond to the Gene Ontology IDs given, and the last column
#'    corresponds to the cell type's marker that led to the result.
#'
#' @seealso \url{https://www.wikidata.org/} for information regarding
#'   Wikidata and \url{https://panglaodb.se/index.html} for the
#'   original database this information was adapted from.
#'
#' @export
#'
#' @examples
#' # Return cell types related to 'muscle contraction' and 'metabolism'
#' go2cell(c("GO:0006936", "GO:0008152"))
#'
#' # Return cell types related to 'response to zinc ion'
#' go2cell("GO:0010043")
#' \dontrun{
#' # IDs should use ':' not '_'
#' go2cell("GO_0010043")
#' }
go2cell <- function(go_ids) {
  go_ids_collapsed <- .collapse_as_values(go_ids, quotes = TRUE)

  genes_from_go <- .get_genes_from_go(go_ids_collapsed)

  gene_values <- unique(genes_from_go$gene)

  celltypes <- .query_ctp_turtle(gene_values, query_key = "cell_type", selector = "gene", mode = "go2cell")

  celltypes_collapsed <- .collapse_as_values(celltypes$cell_type)

  celltype_wdt <- .get_celltype_items(celltypes_collapsed)

  final_table <- celltypes %>%
    dplyr::inner_join(genes_from_go,
      by = "gene"
    ) %>%
    dplyr::left_join(celltype_wdt, by = c("cell_type")) %>%
    dplyr::select(
      cell_type, cell_typeLabel, go_ids, go_termLabel,
      geneLabel
    )
}


#' Get GO ids from cell types
#'
#' Return GO identifiers related to gene markers
#' of specific cell types.
#'
#' This function takes in a character vector of
#' cell type Q identifiers (QIDs), queries Wikidata
#' (\url{https://www.wikidata.org/})
#' and a local RDF database to return GO IDs related
#' to gene markers of these cell types.
#'
#' @param celltype_qids A character vector of cell type Wikidata IDs
#'
#' @return A dataframe of 5 columns: The first two correspond to cell types,
#'    that is, their Wikidata Identifier and their name. The two following
#'    columns correspond to the Gene Ontology IDs given, and the last column
#'    corresponds to the cell type's marker that led to the result.
#'
#' @seealso \url{https://www.wikidata.org/} for information regarding
#'   Wikidata and \url{https://panglaodb.se/index.html} for the
#'   original database this information was adapted from. For an
#'   example of a cell type item in Wikidata, see
#'   \url{https://www.wikidata.org/wiki/Q101405098}
#'
#' @export
#'
#' @examples
#' # Return GO IDs related to cell type
#' # 'human smooth muscle cell'
#' cell2go("Q101404901")
#' \dontrun{
#' # IDs should always start with 'Q'
#' go2cell("101404901")
#' }
cell2go <- function(celltype_qids) {
  if (any(grepl("Q\\d+", celltype_qids) == FALSE)) {
    stop("A QID given doesn't match the regex 'Q\\d+',
             check your input for inconsistencies.")
  }

  with_wd <- stringr::str_glue("wd:{celltype_qids}")

  celltypes <- .query_ctp_turtle(with_wd, query_key = "gene", selector = "cell_type", mode = "cell2go")

  gene_values <- .collapse_as_values(celltypes$gene)

  go_from_genes <- .get_go_from_genes(gene_values)

  celltypes_collapsed <- .collapse_as_values(celltypes$cell_type)

  celltype_wdt <- .get_celltype_items(celltypes_collapsed)

  final_table <- celltypes %>%
    dplyr::inner_join(go_from_genes, by = "gene") %>%
    dplyr::left_join(celltype_wdt, by = c("cell_type")) %>%
    dplyr::select(
      cell_type, cell_typeLabel, go_term, go_itemLabel,
      geneLabel
    )
}
